### ISSUE TITLE
Question: Confusion around keys/access policies

### ISSUE THREAD

**Issue by j--w** [binoculars/awsm-s3tokenvendor#6]
I was playing around trying to get this working tonight and was able to get it to return correct data as far as I can tell, but when I go to upload the file using the key, signature and policy that the endpoint returns, I get the error "The AWS Access Key Id you provided does not exist in our records."

I'm assuming I'm missing something in access policy or configuration for the lambda function or role but I haven't found much help in the docs. Any ideas?

**Comment by binoculars** [comment]
@j--w I'm definitely going to look into this over the weekend. Could you provide the code you are using to do the upload on the browser side (at least the relevant portion)?

**Comment by j--w** [comment]
Thanks @binoculars. I've created a gist with the pertinent information, linked below. For now all I'm doing is sending an example request with postman to get the signature with the filename and size, then I'm manually copying the values from that response info into an html form on localhost to test. I post to my s3 bucket endpoint and get the error as outlined in the gist. Let me know if you need more info or if theres something obvious I've missed. Cheers. Once I figure this out I can submit a PR with a getting-started.md or something to help anyone else who may be checking this out.
https://gist.github.com/j--w/e41db4761ecc144b5366

**Comment by binoculars** [comment]
@j--w Thanks for the gist. I did look into it this weekend and I believe I know what the issue is now. I have something similar (pre-JAWS) that I created a separate IAM user for and then gave it S3:PutObject permissions to the bucket. Apparently the API gateway invokes the lambda function with an ASIAXXXXX access key (vice AKIAXXXXX) which I believe means that it uses the additional security token.

TL;DR I think we need to pass the security token or create an IAM user with bucket permissions. I will try the former and if it works, update the module.

**Comment by j--w** [comment]
Ahh yes, that makes complete sense. I think creating a user with bucket permissions and passing it along makes sense, it allows security concerns to be separated.

**Comment by binoculars** [comment]
I tried passing the session token and that's not working. I suppose the _only_ way ahead is creating the IAM user. I'm going to have to figure out how to do this in CloudFormation that captures the Access Key ID and the Secret so that the function can read them in. Suggestions welcome.

Btw--I know how to do this manually, but that's no fun and kind of defeats the purpose of JAWS.

**Comment by j--w** [comment]
Looks like it may be able to be done with:
http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-iam-user.html
and http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-getatt.html

I'm likely overlooking something, I'll try to dig in a bit more detail.

Is the manual workaround to manually create an IAM user with the right policy and include the access key and secret in environment variables that get included as part of the AWSM module?

**Comment by binoculars** [comment]
Yes, although you have to rename them if you're going to include them in process.env so that the STS stuff doesn't get overwritten. I prefer to keep them in a JSON file and just do `var creds = require('creds.json');`, where creds.json is something like:

``` JSON
{
    "AWS_ACCESS_KEY_ID": "AKID...",
    "AWS_SECRET_ACCESS_KEY": "secret..."
}
```

**Comment by j--w** [comment]
Right, that probably works for my purposes but having cloudformation provision it automatically would be awesome. One of my coworkers is looking into JAWS/cloudformation heavily right now so I'm going to plant the seed to keep this in mind as he's working with it :smile: Paging @dijitalmunky

**Comment by binoculars** [comment]
I'm using this for testing the front end: https://gist.github.com/binoculars/e8b68e6b4e1c2e5f42ba
